File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    vector<int> answer;
    map<char, pair<int, int>> way;
    
    way['N'] = {0, -1};
    way['S'] = {0, 1};
    way['W'] = {-1, 0};
    way['E'] = {1, 0};
    
    int x = 0, y = 0;
    int h = park.size(), w = park[0].size();
    
    
    for (int i = 0; i < h; i++){
        for (int j = 0; j < w; j++){
            if (park[i][j] == 'S'){
                x = j;
                y = i;
            }
        }
    }
    
    cout << y << " " << x << endl;
    
    for (string route: routes){
        
        int nx = x;
        int ny = y;
        bool can = true;
        
        for (int i = 0; i < route[2] - '0'; i++){
             nx += way[route[0]].first;
             ny += way[route[0]].second;
            
            if (nx < 0 || ny < 0 || nx >= w || ny >= h || park[ny][nx] == 'X'){
                can = false;
                break;
            }
            cout << ny << "," << nx << "  ";
        }
        cout << can << "      ";
        cout << endl;
        
        if (can){
            x = nx;
            y = ny;
        }
    }
    
    return {y, x};
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    n, m = len(park), len(park[0])
    ways = {"W":[-1, 0], "N":[0, -1], "S":[0, 1], "E":[1, 0]}
    s = [0, 0]
    
    for i in range(n):
        for j in range(m):
            if park[i][j] == 'S':
                s[0] = j
                s[1] = i
                break
    
    x, y = s[0], s[1]
    print(y, x)
    for route in routes:
        w, cnt = route.split(' ')
        cnt = int(cnt)
        
        nx, ny = x, y
        ok = True
        for i in range(cnt):
            nx += ways[w][0]
            ny += ways[w][1]
            
            if nx < 0 or ny < 0 or nx >= m or ny >= n or park[ny][nx] == 'X':
                ok = False
                break
            
        if ok:
            x = nx
            y = ny
        print(y, x)
        
    return [y, x]